lib/web/apiserver_test.go

```diff
@@ -2002,6 +2003,106 @@ func TestGetMFADevices(t *testing.T) {
 	require.Equal(t, "otp", devices[0].Name) // default device name
 }
 
+func TestCreateAuthenticateChallenge_WithAuth(t *testing.T) {
+	t.Parallel()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	// Creates a user with a TOTP device, with second factor preference to OTP only.
+	pack := proxy.authPack(t, "llama@example.com")
+
+	endpoint := pack.clt.Endpoint("webapi", "mfa", "authnchallenge")
+	re, err := pack.clt.PostJSON(context.Background(), endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var chal auth.MFAAuthenticateChallenge
+	err = json.Unmarshal(re.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
+}
+
+func TestCreateAuthenticateChallenge_WithToken(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	ap, err := types.NewAuthPreference(types.AuthPreferenceSpecV2{
+		Type:         constants.Local,
+		SecondFactor: constants.SecondFactorOTP,
+	})
+	require.NoError(t, err)
+	err = env.server.Auth().SetAuthPreference(ctx, ap)
+	require.NoError(t, err)
+
+	// Create a user with a TOTP device.
+	username := "llama"
+	proxy.createUser(ctx, t, username, "root", "password", "some-otp-secret")
+
+	// Acquire a start token.
+	startToken, err := types.NewUserToken("some-token-id")
+	require.NoError(t, err)
+	startToken.SetUser(username)
+	startToken.SetSubKind(auth.UserTokenTypeRecoveryStart)
+	startToken.SetExpiry(env.clock.Now().Add(5 * time.Minute))
+	_, err = env.server.Auth().Identity.CreateUserToken(ctx, startToken)
+	require.NoError(t, err)
+
+	// Call the endpoint.
+	clt := proxy.newClient(t)
+	endpoint := clt.Endpoint("webapi", "mfa", "token", startToken.GetName(), "authnchallenge")
+	res, err := clt.PostJSON(ctx, endpoint, url.Values{})
+	require.NoError(t, err)
+
+	var chal auth.MFAAuthenticateChallenge
+	err = json.Unmarshal(res.Bytes(), &chal)
+	require.NoError(t, err)
+	require.True(t, chal.TOTPChallenge)
+	require.Empty(t, chal.U2FChallenges)
+	require.Empty(t, chal.WebauthnChallenge)
+}
+
+func TestCreateAuthenticateChallenge_WithUserCreds(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+	env := newWebPack(t, 1)
+	proxy := env.proxies[0]
+
+	// Create a user with a TOTP device, with second factor preference to OTP only.
+	pack := proxy.authPack(t, "llama@example.com")
+
+	requestBody := client.MFAChallengeRequest{
+		User: pack.user,
+		Pass: pack.password,
+	}
+
+	// Test private endpoint.
+	endpoint := pack.clt.Endpoint("webapi", "u2f", "password", "changerequest")
```

Comment:
Should this be `/webapi/mfa/authnchallenge` ?

Reply:
no, this endpoint also calls CreateAuthenticateChallenge but is used for re-auth second factor when changing password. i changed the test to your genius suggestion to run it in a loop, so hopefully the test makes more sense

Reply:
Yeah, I thought I might be missing something. Thanks for clarifying.